### DISCUSSION TITLE
TinyUSB memory usage question

### DISCUSSION THREAD

**Discussion by Unturned3** [hathach/tinyusb#2548] [General]
In the [cdc_msc_freertos](https://github.com/hathach/tinyusb/blob/35aa9d06e141855c39cb50be4d15308ee11f3489/examples/device/cdc_msc_freertos/src/main.c#L42) example, the `usb_device_task` is given 4096 _words_ (i.e. 16KB) for its stack. Is this just an aribtrary value, or the minimum size needed for TinyUSB to function? I was quite surprised by how much RAM the task needs.

In contrast, ST's USB device CDC implementation (the one that comes with CubeIDE) seems to take only a few KBs of memory.

**Comment by hathach** [comment]
4096 is given for esp32 only since it requires a bit for logging/printf and esp32 has plenty of ram. For stm32 you can try with 256/512 to see if that works. Stack size could depends on dynamic event and workflow, so you still need to monitor its usage with debugger/sysview or simply printf to be sure

**Reply by Unturned3** [reply]
I see, thanks! I tried 256 for the task stack size, and it seemed to work smoothly.

Unfortunately STM32F411CE only has 4 USB endpoints... took me quite some head-banging to eventually realize why [cdc_dual_ports](https://github.com/hathach/tinyusb/tree/master/examples/device/cdc_dual_ports) was not working. I thought something was wrong with my FreeRTOS setup.
